### PR TITLE
Add server-side formspec context storage

### LINKED ISSUE #7535: Add server-side secure formspec storage
Add the ["context"](https://rubenwardy.com/minetest_modding_book/en/chapters/formspecs.html#contexts) pattern to the engine - allow storing some data between formspec sending and receive_fields

From #7391

### PR DESCRIPTION
**Pull request by rubenwardy**
Fixes #7535. Allows mods to store arbitrary Lua information to be passed to on_receive_fields callbacks.

Note: this is Lua API level, and technically not to do with formspecs

### CODE REVIEW

**Comment by rubenwardy** [comment]
This is a bad idea for a few reasons:

* It's essentially a race condition - there's only one table to store the contexts. Say mod A has a formspec open, mod B shows a formspec, then the server receives the fields from mod A. This would result in the wrong context being given
* It's added complexity for something that is already pretty easy

File: src/script/cpp_api/s_player.cpp [hunk 1]
```diff
@@ -213,7 +215,19 @@ void ScriptApiPlayer::on_playerReceiveFields(ServerActiveObject *player,
...
 	}
-	runCallbacks(3, RUN_CALLBACKS_MODE_OR_SC);
+	// param 4
+	if (formname.empty()) {
+		lua_pushnil(L);
+	} else {
+		lua_getglobal(L, "core");
+		int tmp = lua_gettop(L);
+		lua_getfield(L, -1, "fs_contexts");
+		lua_getfield(L, -1, ((PlayerSAO*)player)->getPlayer()->getName());
```
Comment (DS-Minetest) [inline src/script/cpp_api/s_player.cpp]:
```suggestion
		lua_getfield(L, -1, ((PlayerSAO *)player)->getPlayer()->getName());
```

File: src/script/cpp_api/s_player.cpp [hunk 2]
```diff
@@ -213,7 +215,19 @@ void ScriptApiPlayer::on_playerReceiveFields(ServerActiveObject *player,
 		lua_pushlstring(L, value.c_str(), value.size());
 		lua_settable(L, -3);
 	}
-	runCallbacks(3, RUN_CALLBACKS_MODE_OR_SC);
+	// param 4
+	if (formname.empty()) {
+		lua_pushnil(L);
+	} else {
+		lua_getglobal(L, "core");
+		int tmp = lua_gettop(L);
```
Comment (SmallJoker) [inline src/script/cpp_api/s_player.cpp]: unused variable

File: src/script/lua_api/l_server.cpp [hunk 3]
```diff
@@ -360,12 +360,16 @@ int ModApiServer::l_show_formspec(lua_State *L)
...
 	const char *formspec = luaL_checkstring(L, 3);
 
-	if(getServer(L)->showFormspec(playername,formspec,formname))
-	{
-		lua_pushboolean(L, true);
-	}else{
-		lua_pushboolean(L, false);
-	}
+	lua_getglobal(L, "core");
+	lua_getfield(L, -1, "fs_contexts");
```
Comment (SmallJoker) [inline src/script/lua_api/l_server.cpp]: undocumented field

**Comment by SmallJoker** [comment]
> It's added complexity for something that is already pretty easy

Indeed it is, and it would be the same field for all mods, which might cause conflicts if one doesn't handle it carefully. Per-formspec name callbacks could solve that.

**Pull request closed without merging**